python/lsst/ts/salobj/domain.py

```diff
@@ -0,0 +1,250 @@
+                                                     kind=dds.DDSHistoryKind.KEEP_LAST)
+            self.reader_qos = self.qos_provider.get_reader_qos()
+            self.reader_qos.set_policies([read_queue_policy])
+            """Quality of service for topic readers that want late joiner data.
+
+            A dds.Qos.
+            """
+
+            self.volatile_reader_qos = self.qos_provider.get_reader_qos()
+            volatile_policy = dds.DurabilityQosPolicy(dds.DDSDurabilityKind.VOLATILE)
+            self.volatile_reader_qos.set_policies([read_queue_policy, volatile_policy])
+            """Quality of service for topic readers that do not want
+            late joiner data.
+
+            Examples include command_<name> topic readers and the
+            ackcmd topic reader. A dds.Qos.
+            """
+        except Exception:
+            # very unlikely, but just in case...
+            self.participant.close()
+            raise
+
+    def add_salinfo(self, salinfo):
+        """Add the specified salinfo to the internal registry.
+
+        Parameters
+        ----------
+        salinfo : `SalInfo`
+            SAL component information
+
+        Raises
+        ------
+        RuntimeError
+            If the salinfo is already present
+        """
+        if salinfo in self._salinfo_set:
+            raise RuntimeError(f"salinfo {salinfo} already added")
+
+    def remove_salinfo(self, salinfo):
+        """Remove the specified salinfo from the internal registry.
+
+        Parameters
+        ----------
+        salinfo : `SalInfo`
+            SAL component information
+
+        Returns
+        -------
+        removed : `boolean`
+            True if removed, False if not found
+        """
+        try:
+            self._salinfo_set.remove(salinfo)
+            return True
+        except KeyError:
+            return False
+
+    async def close(self):
+        """Close all registered `SalInfo` and the dds domain participant"""
+        if self.participant is None:
+            return
+        while self._salinfo_set:
+            salinfo = self._salinfo_set.pop()
+            await salinfo.close()
+        self.close_dds()
+        # give read loops a bit more time
+        await asyncio.sleep(0.01)
+        if self.num_read_loops != 0 or self.num_read_threads != 0:
+            warnings.warn(f"After Domain.close num_read_loops={self.num_read_loops} and "
+                          f"num_read_threads={self.num_read_threads}; both should be 0")
+
+    def close_dds(self):
+        """Close the dds DomainParticipant."""
+        participant = getattr(self, "participant", None)
+        if participant is not None:
+            participant.close()
+            self.participant = None
+
+    def tai_utc(self, utc):
+        """Return TAI-UTC (seconds) given UTC in unix seconds.
+
+        Warnings
+        --------
+        This is a temporary hack.
+        """
+        return 37
```

Comment:
Maybe add a `#FIXME` tag?

Reply:
Good suggestion. I filed DM-19791 and added a TODO reference to it.